### ISSUE TITLE
Move the low-level parallel upload APIs to the public namespace.

### ISSUE THREAD

**Issue by teo-tsirpanis** [googleapis/google-cloud-cpp#11145]
**What component of `google-cloud-cpp` is this feature request for?**
`google/cloud/storage`

**Is your feature request related to a problem? Please describe.**
I have big in-memory buffers that I want to upload to GCS. [`PrepareParallelUpload`](https://github.com/googleapis/google-cloud-cpp/blob/70081b549bad1d4eea2bc37681ce100db2e75078/google/cloud/storage/doc/parallel-uploads-design.md#prepareparallelupload) seems like a fit to my needs, but unfortunately it is declared in the `google::cloud::storage::internal` namespace, discouraging external code from using it.

**Describe the solution you'd like**
I would like `PrepareParallelUpload` (and `CreateUploadShards`, might be useful for others) to become available in the public `google::cloud::storage` namespace. Currently only `ParallelUploadFile` is public; don't know if it's intentional or an oversight.

**Describe alternatives you've considered**
Using them from the `internal` namespace doesn't sound a good idea.

**Comment by coryan** [comment]
> I have big in-memory buffers

can you share how big?  There may be simpler / more efficient alternatives.

> that I want to upload to GCS. [`PrepareParallelUpload`](https://github.com/googleapis/google-cloud-cpp/blob/70081b549bad1d4eea2bc37681ce100db2e75078/google/cloud/storage/doc/parallel-uploads-design.md#prepareparallelupload) seems like a fit to my needs, 

Maybe. That would use resumable uploads, which are slower than single-shot uploads.  The new non-copying version of `InsertObject()` would allow you to upload those buffers without making copies.  These will appear in the next release:

https://googleapis.dev/cpp/google-cloud-storage/HEAD/classgoogle_1_1cloud_1_1storage_1_1Client.html#ace17975b9eeae9df0da67712a430a349

For in-memory buffers that may be the most efficient way to upload the buffer.  If the buffers are not too big (say up to a few GiB), something like this could work:

```cc
namespace g = google::cloud;
namespace gcs = google::cloud::storage;

// TODO: actually compile and test this.
// TODO: the error handling is a bit naive.
gcs::ObjectMetadata
ParallelUpload(
    gcs::Client client, std::string bucket_name, std::string object_name, absl::string_view data) {
  auto const kShardSize = 32 * 1024 * 1024L;
  auto const prefix = gcs::CreateRandomPrefix();
  struct Shard {
    std::string name;
    absl::string_view data;
  };
  std::vector<Shard> shards;
  shards.reserve(data.size() / kShardSize + 1);
  int counter = 0;
  for (std::size_t offset = 0; offset < data.size(); offset += kShardSize) {
      shards.push_back(Shard{prefix + "/shard-" + counter, Shard{data.substr(offset, kShardSize)});
  }
  std::vector<std::future<g::StatusOr<gcs::ObjectMetadata>>> tasks(shards.size());
  std::transform(shards.begin(), shards.end(), tasks.begin(), [&](auto shard) mutable {
    return std::async(std::launch::async, [](auto client, auto bucket, auto object, auto data) {
        return client.InsertObject(bucket, object, data);
      }, client, bucket_name, shard.name, shard.data);   
  });
  std::vector<gcs::ObjectMetadata> results(tasks.size());
  std::transform(tasks.begin(), tasks.end(), results.begin(), [](auto& f) {
      auto metadata = f.get();
      if (!metadata) throw std::move(metadata).status();
     return *std::move(metadata);
  });
  std::vector<gcs::ComposeSourceObject> sources(results.size());
  std::transform(results.begin(), results.end(), sources.begin(), [](auto m) {
     return ComposeSourceObject{m.name(), m.generation(), absl::nullopt};
  });
  auto metadata = gcs::ComposeMany(
      client, bucket_name, sources, prefix, object_name, /*ignore_cleanup_failures=*/true);
  if (!metadata) throw std::move(metadata).status();
  auto status = gcs::DeleteByPrefix(client, bucket_name, prefix);
  if (!status.ok()) throw status;
  return *std::move(metadata);
}
```

For much larger buffers (say starting at around 4 GiB) you may want to create fewer shards, or use resumable uploads (to avoid resending data on partial failures).

> but unfortunately it is declared in the `google::cloud::storage::internal` namespace, discouraging external code from using it.

Correct.

> **Describe the solution you'd like** I would like `PrepareParallelUpload` (and `CreateUploadShards`, might be useful for others) to become available in the public `google::cloud::storage` namespace. Currently only `ParallelUploadFile` is public; don't know if it's intentional or an oversight.

It is intentional, we did not know if the API was useful or if we wanted to tie the implementation down by exposing the implementation details.  For example, we could change the implementation to use single-shot uploads (when the shards are small enough), or to use XML multipart uploads, or to use the asynchronous gRPC client (which is progressing too slowly), and then this function would disappear or its interface would completely change.

> Using them from the `internal` namespace doesn't sound a good idea.

Certainly not recommended.

**Comment by ihnorton** [comment]
> can you share how big?

Size varies, but can be 10s of GB per object in some cases. We currently [use the compose API](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/filesystem/gcs.cc#L996-L1003) for parallel uploads. We recently looked at the new (to us) multipart support, and are evaluating replacing compose with multipart, because compose has some drawbacks -- at least relative to the S3 multipart semantics we're familiar with: the 32 item limit requires tiered composition, and compose objects are individually visible, so we needed temp naming.

But we see multipart is [not supported yet in this SDK](https://github.com/googleapis/google-cloud-cpp/issues/7054), as well as the doc comment that `Cloud Storage can take several minutes to process a request to complete a multipart upload.` ([source](https://cloud.google.com/storage/docs/xml-api/post-object-complete)).

In the course of further investigation, @teo-tsirpanis observed that resumable uploads are recommended in the docs for large files, but we weren’t yet able to figure out how to do parallel resumable uploads with the SDK's public API (whereas it appears to be possible with the REST API and internal code).

>  The new non-copying version of `InsertObject()` would allow you to upload those buffers without making copies. 
> …
> For in-memory buffers that may be the most efficient way to upload the buffer.

That looks very useful. Will InsertObject perform parallel uploads, and if so, is the size threshold or the level of parallelism controllable? Also, is there any expected timeframe for a release with the new `InsertObject` API?

Thanks.

**Comment by coryan** [comment]
> > can you share how big?
> 
> Size varies, but can be 10s of GB per object in some cases. We currently [use the compose API](https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/filesystem/gcs.cc#L996-L1003) for parallel uploads. We recently looked at the new (to us) multipart support, and are evaluating replacing compose with multipart, because compose has some drawbacks -- at least relative to the S3 multipart semantics we're familiar with: the 32 item limit requires tiered composition, and compose objects are individually visible, so we needed temp naming.

Yes, these are understood problems with the compose API.  Some people use a separate bucket to avoid making the components visible in the man bucket, and then use `RewriteObject()` to make a copy.  If the source and target buckets are in the same location and storage class these rewrites are very efficient (effectively they are just metadata changes in the service).

Using a separate bucket also allows you to set a lifecycle policy to automatically garbage collect the temporary objects and components.

> In the course of further investigation, @teo-tsirpanis observed that resumable uploads are recommended in the docs for large files, but we weren’t yet able to figure out how to do parallel resumable uploads with the SDK's public API (whereas it appears to be possible with the REST API and internal code).

This is all correct.

> 
> > The new non-copying version of `InsertObject()` would allow you to upload those buffers without making copies.
> > …
> > For in-memory buffers that may be the most efficient way to upload the buffer.
> 
> That looks very useful. Will InsertObject perform parallel uploads,

No, but that sounds like an interesting idea.  Maybe not for the function itself, but I can see us turning that code I wrote above into a `google::cloud::storage::ParallelInsertObject()` function.

> and if so, is the size threshold or the level of parallelism controllable?

If/When we write such a function, we will probably make the parallelism controllable.

> Also, is there any expected timeframe for a release with the new `InsertObject` API?

Next week.  We create releases at the beginning of each month.  In this case we will start the release process on 2023-04-03.  This normally takes a few hours, but it could be take a couple of days if we run into problems.

**Comment by ihnorton** [comment]
Thanks @coryan. Given:
> Yes, these are understood problems with the compose API.
> I can see us turning that code I wrote above into a google::cloud::storage::ParallelInsertObject() function.

IIUC that code sample you shared is using Compose, right? Would it make sense to implement ParallelInsertObject in terms of parallel resumable uploads instead?

Fundamentally, I'm still trying to understand whether there is a reason _not_ to use parallel resumable uploads as a replacement for Compose. If resumable uploads are not subject to the same limitations/problems as the Compose API, are there some other limitations that make it unsuitable for this use-case?

Thanks

**Comment by coryan** [comment]
> IIUC that code sample you shared is using Compose, right? 

Yes.

> Would it make sense to implement ParallelInsertObject in terms of parallel resumable uploads instead?

I am not sure exactly what do you have in mind when you say "parallel resumable uploads".   If this refers to the internal API that started this thread ([PrepareParallelUpload](https://github.com/googleapis/google-cloud-cpp/blob/70081b549bad1d4eea2bc37681ce100db2e75078/google/cloud/storage/doc/parallel-uploads-design.md#prepareparallelupload)): that creates multiple resumable uploads but still needs to perform a Compose at the end to create a single object.  If you have something else in mind, then I got lost somewhere.

I chose to use `InsertObject()` because that is a smidge more efficient if you have all the data in memory: in the common case a resumable upload requires at least two roundtrips to the service.  `InsertObject()` use a single-shot (non-resumable) upload and requires a single roundtrip to the service (in the common case).

> Fundamentally, I'm still trying to understand whether there is a reason _not_ to use parallel resumable uploads as a replacement for Compose.

There is no replacement for Compose in the JSON API.  If you want to parallelize an upload that is what you need to do at the end.  The XML API does offer MPU (Multi-Part Uploads), but at the moment none of the client libraries for GCS, including the C++ client, uses the XML API and consequently none supports XML MPUs.

>  If resumable uploads are not subject to the same limitations/problems as the Compose API, are there some other limitations that make it unsuitable for this use-case?

See above.  I think either you misunderstand what these parallel uploads do, or I misunderstand what you mean by them.

**Comment by ihnorton** [comment]
> I think either you misunderstand what these parallel uploads do

You're right, I think I've misunderstood the ["Multiple chunk upload JSON API"](https://cloud.google.com/storage/docs/performing-resumable-uploads#json-api_2) section of the docs.

I assumed that this API allowed different chunks to be sent in parallel, but based on your responses (and [this feature request](https://issuetracker.google.com/issues/220523936?pli=1), which I did not see until now), that assumption seems wrong.

> The XML API does offer MPU (Multi-Part Uploads)

Thanks, have been looking at that, but the warning in the docs about `several minutes to process a request to complete a multipart upload` is concerning.

Thanks for the responses here. Big +1 to the `ParallelInsertObject` proposal, especially if that could seamlessly support "parallel resumable uploads" as an option, if that becomes supported on the backend side in the future. (Please feel free to close or re-title this as you see fit)

**Comment by coryan** [comment]
Closing in favor of #11191.
